cpp/src/arrow/util/decimal.cc

```diff
@@ -873,4 +874,114 @@ Status Decimal128::Rescale(int32_t original_scale, int32_t new_scale,
   return Status::OK();
 }
 
+Status Decimal128::FromBigEndian(const uint8_t* bytes, int32_t length, Decimal128* out) {
+  static constexpr int32_t kMinDecimalBytes = 1;
+  static constexpr int32_t kMaxDecimalBytes = 16;
+
+  int64_t high;
+  uint64_t low;
+
+  if (length < kMinDecimalBytes || length > kMaxDecimalBytes) {
+    std::ostringstream stream;
+    stream << "Length of byte array passed to Decimal128::FromBigEndian ";
+    stream << "was " << length << ", but must be between ";
+    stream << kMinDecimalBytes << " and " << kMaxDecimalBytes;
+    Status::Invalid(stream.str());
+  }
+
+  /// Bytes are coming in big-endian, so the first byte is the MSB and therefore holds the
+  /// sign bit.
+  const bool is_negative = static_cast<int8_t>(bytes[0]) < 0;
+
+  /// Sign extend the low bits if necessary
+  low = UINT64_MAX * (is_negative && length < 8);
+  high = -1 * (is_negative && length < kMaxDecimalBytes);
+
+  /// Stop byte of the high bytes
+  const int32_t high_bits_offset = std::max(0, length - 8);
+
+  /// Shift left enough bits to make room for the incoming int64_t
+  high <<= high_bits_offset * CHAR_BIT;
+
+  /// Preserve the upper bits by inplace OR-ing the int64_t
+  uint64_t value;
+  RETURN_NOT_OK(BytesToInteger(bytes, high_bits_offset, &value));
+  high |= value;
+
+  /// Stop byte of the low bytes
+  const int32_t low_bits_offset = std::min(length, 8);
+
+  /// Shift left enough bits to make room for the incoming uint64_t
+  low <<= low_bits_offset * CHAR_BIT;
+
+  /// Preserve the upper bits by inplace OR-ing the uint64_t
+  RETURN_NOT_OK(
+      BytesToInteger(bytes + high_bits_offset, length - high_bits_offset, &value));
+  low |= value;
+
+  *out = Decimal128(high, low);
+  return Status::OK();
+}
+
+Status Decimal128::BytesToInteger(const uint8_t* bytes, int32_t length, uint64_t* out) {
+  using ::arrow::BitUtil::FromBigEndian;
+
+  switch (length) {
+    case 0:
+      *out = 0;
+      break;
+    case 1:
+      *out = *bytes;
+      break;
+    case 2:
+      *out = BitUtil::FromBigEndian(*reinterpret_cast<const uint16_t*>(bytes));
```

Comment:
This is potentially doing an unaligned access, which is forbidden on some architectures (SPARC for example, perhaps others).

Reply:
Looks like for linux there's [`get_unaligned`/`put_unaligned`](https://www.kernel.org/doc/Documentation/unaligned-memory-access.txt). For MSVC, we could create a template function that creates a temporary struct and access a member.

Reply:
I don't think we need such complication. Just read the memory byte per byte. Or, if you really insist on doing a single `int64_t` read, use `mempcy` into an `int64_t` local variable. Most compilers will optimize that into an unaligned access if the platform allows for it.

Reply:
I looked at the generated assembly using the switch statement and it's producing a series of cmp/jne statements, so implementing it in a loop will have the same performance without the alignment issues. I will make the change and push it.